From a17573dd3b2dc3787c4df8bcf7c95a954abde37d Mon Sep 17 00:00:00 2001
From: CI BOT <CIBOT@noreply.com>
Date: Fri, 17 Jul 2026 10:23:19 +0200
Subject: [PATCH] Cherry pick support for LLVM 22

Cherry pick from apache/tvm main branch commit:
* 24a4e6cb [LLVM] Fix build failures when building with llvm>=22 (#18818)

Fixes build failures when compiling against LLVM >= 22
(TVM_LLVM_VERSION >= 220):

- llvm::cl::getRegisteredOptions() returns a
  llvm::DenseMap<llvm::StringRef, llvm::cl::Option*> instead
  of llvm::StringMap<llvm::cl::Option*>&, so lookups now use
  find()/end() instead of count()/operator[].
- llvm::TargetOptions no longer has an UnsafeFPMath member.
- llvm::Target::createTargetMachine() and
  llvm::TargetRegistry::lookupTarget() now take a
  llvm::Triple instead of a triple string.
---
 src/target/llvm/llvm_instance.cc | 54 +++++++++++++++++++++++++++++---
 1 file changed, 49 insertions(+), 5 deletions(-)

diff --git a/src/target/llvm/llvm_instance.cc b/src/target/llvm/llvm_instance.cc
index e494a2bbf..23efe66a9 100644
--- a/src/target/llvm/llvm_instance.cc
+++ b/src/target/llvm/llvm_instance.cc
@@ -232,7 +232,7 @@ LLVMTargetInfo::LLVMTargetInfo(LLVMInstance& instance, const TargetJSON& target)
   }
 
   if (const auto& v = Downcast<Optional<Array<String>>>(target.Get("cl-opt"))) {
-    llvm::StringMap<llvm::cl::Option*>& options = llvm::cl::getRegisteredOptions();
+    auto& options = llvm::cl::getRegisteredOptions();
     bool parse_error = false;
     for (const String& s : v.value()) {
       Option opt = ParseOptionString(s);
@@ -240,7 +240,11 @@ LLVMTargetInfo::LLVMTargetInfo(LLVMInstance& instance, const TargetJSON& target)
         parse_error = true;
         continue;
       }
+#if TVM_LLVM_VERSION >= 220
+      if (options.find(opt.name) != options.end()) {
+#else
       if (options.count(opt.name)) {
+#endif
         llvm_options_.push_back(opt);
       } else {
         // Flag an error, but don't abort. LLVM flags may change, and this would
@@ -315,7 +319,9 @@ LLVMTargetInfo::LLVMTargetInfo(LLVMInstance& instance, const TargetJSON& target)
   // In clang, these are fed from LangOpts which describe language specific features
   // TODO(AndrewZhaoLuo): figure out how these relate to fast math flags
   target_options_.AllowFPOpFusion = llvm::FPOpFusion::Fast;
+#if TVM_LLVM_VERSION < 220
   target_options_.UnsafeFPMath = false;
+#endif
   target_options_.NoInfsFPMath = false;
   target_options_.NoNaNsFPMath = true;
   target_options_.FloatABIType = float_abi;
@@ -411,9 +417,16 @@ LLVMTargetInfo::~LLVMTargetInfo() = default;
 static const llvm::Target* CreateLLVMTargetInstance(const std::string triple,
                                                     const bool allow_missing = true) {
   std::string error;
+#if TVM_LLVM_VERSION >= 220
+  llvm::Triple triple_obj(triple);
+#endif
   // create LLVM instance
   // required mimimum: llvm::InitializeAllTargets()
+#if TVM_LLVM_VERSION >= 220
+  const llvm::Target* llvm_instance = llvm::TargetRegistry::lookupTarget(triple_obj, error);
+#else
   const llvm::Target* llvm_instance = llvm::TargetRegistry::lookupTarget(triple, error);
+#endif
   if (!allow_missing && !llvm_instance) {
     ICHECK(llvm_instance) << "LLVM instance error: `" << error << "`";
   }
@@ -431,8 +444,14 @@ static std::unique_ptr<llvm::TargetMachine> CreateLLVMTargetMachine(
 #else
     const llvm::CodeGenOptLevel& opt_level = llvm::CodeGenOptLevel(0)) {
 #endif
+#if TVM_LLVM_VERSION >= 220
+  llvm::Triple triple_obj(triple);
+  llvm::TargetMachine* tm = llvm_instance->createTargetMachine(
+      triple_obj, cpu, features, target_options, reloc_model, code_model, opt_level);
+#else
   llvm::TargetMachine* tm = llvm_instance->createTargetMachine(
       triple, cpu, features, target_options, reloc_model, code_model, opt_level);
+#endif
   ICHECK(tm != nullptr);
 
   return std::unique_ptr<llvm::TargetMachine>(tm);
@@ -813,8 +832,22 @@ bool LLVMTargetInfo::MatchesGlobalState() const {
 }
 
 void LLVMTargetInfo::GetOptionValue(LLVMTargetInfo::Option* opt) const {
-  llvm::StringMap<llvm::cl::Option*>& options = llvm::cl::getRegisteredOptions();
-  llvm::cl::Option* base_op = options[opt->name];
+  auto& options = llvm::cl::getRegisteredOptions();
+  llvm::cl::Option* base_op = nullptr;
+#if TVM_LLVM_VERSION >= 220
+  auto it = options.find(opt->name);
+  if (it != options.end()) {
+    base_op = it->second;
+  }
+#else
+  if (options.count(opt->name)) {
+    base_op = options[opt->name];
+  }
+#endif
+  if (base_op == nullptr) {
+    opt->type = Option::OptType::Invalid;
+    return;
+  }
 
   if (opt->type == Option::OptType::Bool) {
     auto* bool_op = static_cast<llvm::cl::opt<bool>*>(base_op);
@@ -994,7 +1027,7 @@ void LLVMTarget::SetTargetMetadata(llvm::Module* module) const {
 }
 
 bool LLVMTarget::ApplyLLVMOptions(bool apply_otherwise_revert, bool dry_run) {
-  llvm::StringMap<llvm::cl::Option*>& options = llvm::cl::getRegisteredOptions();
+  auto& options = llvm::cl::getRegisteredOptions();
   bool changed = false;
 
 #define HANDLE_OPTION_VALUE(option, new_val, saved_val)                  \
@@ -1014,7 +1047,18 @@ bool LLVMTarget::ApplyLLVMOptions(bool apply_otherwise_revert, bool dry_run) {
     const Option& new_opt = new_options[i];
     const Option& saved_opt = saved_llvm_options_[i];
 
-    llvm::cl::Option* base_op = options[new_opt.name];
+    llvm::cl::Option* base_op = nullptr;
+#if TVM_LLVM_VERSION >= 220
+    auto it = options.find(new_opt.name);
+    if (it != options.end()) {
+      base_op = it->second;
+    }
+#else
+    if (options.count(new_opt.name)) {
+      base_op = options[new_opt.name];
+    }
+#endif
+    ICHECK(base_op != nullptr) << "LLVM option not found: " << new_opt.name;
 
     if (new_opt.type == Option::OptType::Bool) {
       auto* bool_op = static_cast<llvm::cl::opt<bool>*>(base_op);
-- 
2.43.0

